connectors/services/attachments.py

```diff
@@ -0,0 +1,250 @@
+        for item in os.scandir(drop_directory):
+            try:
+                if not item.is_file():
+                    continue
+                current_size += d.stat().st_size
+            except Exception:
+                pass
+
+        return current_size <= self.max_disk_size
+
+    def _sanitize(self, name):
+        return SANITIZER.sub("-", name)
+
+    async def drop_file(self, gen, name, filename, index, doc_id):
+        """Writes a file by chunks using the async generator and then a desc file
+
+        When the desc file hits the disk, it'll be picked up by AttachmentService.
+        """
+        if not self.can_drop(self.drop_dir):
+            raise OSError(f"Limit of {self.max_disk_size} bytes reached")
+
+        key = f"{index}-{doc_id}-{self._sanitize(filename)}"
+        target = os.path.join(self.drop_dir, key)
+        async with aiofiles.open(target, "wb") as f:
+            async for chunk in gen:
+                await f.write(chunk)
+
+        logger.info(f"Dropped {key}")
+
+        # the file is now on disk, let's create the desc
+        desc = {
+            # XXX maybe its overkill for v1 since a node only connects to a
+            # single ES
+            # XXX add suport for API key
+            "filename": key,
+            "index": index,
+            # XXX how do we know this value when the initial sync of the doc is
+            # not done yet. we might need to use the source id
+            # and at sync time, query for the corresponding ES doc id
+            # and send it only once it's there..
+            "doc_id": doc_id,
+            "name": filename,
+        }
+
+        desc_file = os.path.join(self.drop_dir, os.path.splitext(key)[0] + ".json")
+        with open(desc_file, "w") as f:
+            f.write(json.dumps(desc))
+
+        logger.info(f"Dropped {desc_file}")
+
+
+class AttachmentService(BaseService):
+    """Watches a directory for files and sends them by chunks
+
+    Uses `Transfer-Encoding: chunked`
+    """
+
+    def __init__(self, args):
+        super().__init__(args)
+        self.elastic_config = self.config["elasticsearch"]
+        self._config = self.config.get("attachments", {})
+        self.max_concurrency = self._config.get("max_concurrency", 5)
+        self.directory = self._prepare_dir("drop")
+        self.idle_time = self._config.get("idling", 30)
+        self.log_directory = self._prepare_dir("logs")
+        self.results_directory = self._prepare_dir("results")
+        self.running = False
+        self._sleeps = CancellableSleeps()
+
+    def _to_b64(self, filename):
+        """Calls the system base64 utility to create a base64-encoded file
+
+        The base64 utility is a stream encoder, the file will not be fully
+        loaded into memory.
+
+        This is a blocking method.
+        """
+        cmd = f"{BASE64} {filename} > {filename}.b64"
+        logger.debug(f"[{os.path.basename(filename)}] Running {cmd}")
+        os.system(cmd)
```

Comment:
Doesn't asyncio support spawning a subprocess?

Reply:
yes that was just for the draft POC

Reply:
`run_in_executor` along with a `ProcessPoolExecutor` see https://pymotw.com/3/asyncio/executors.html